src/main/java/org/jitsi/videobridge/Endpoint.java

```diff
@@ -525,59 +537,92 @@ private void onPinnedEndpointChangedEvent(
+        pinnedEndpointsChanged(Collections.singletonList(newPinnedEndpointID));
+    }
 
-        Endpoint newPinnedEndpoint;
-        if (!StringUtils.isNullOrEmpty(newPinnedEndpointID)
-                && conference != null)
+    /**
+     * Notifies this {@code Endpoint} that a {@code PinnedEndpointsChangedEvent}
+     * has been received by the associated {@code SctpConnection}.
+     *
+     * @param src the {@code WebRtcDataStream} by which {@code jsonObject} has
+     * been received
+     * @param jsonObject the JSON object with {@link Videobridge#COLIBRI_CLASS}
+     * {@code PinnedEndpointChangedEvent} which has been received by the
+     * associated {@code SctpConnection}
+     */
+    private void onPinnedEndpointsChangedEvent(
+            WebRtcDataStream src,
+            JSONObject jsonObject)
+    {
+        // Find the new pinned endpoint.
+        Object o = jsonObject.get("pinnedEndpoints");
+        if (!(o instanceof JSONArray))
         {
-            newPinnedEndpoint = conference.getEndpoint(newPinnedEndpointID);
+            logger.warn("Received invalid or unexpected JSON: " + jsonObject);
+            return;
         }
-        else
+
+        JSONArray jsonArray = (JSONArray) o;
+        List<String> newPinnedEndpoints = new LinkedList<>();
+        for (Object endpointId : jsonArray)
+        {
+            if (endpointId != null && endpointId instanceof String)
+            {
+                newPinnedEndpoints.add((String)endpointId);
+            }
+        }
+
+        if (logger.isDebugEnabled())
         {
-            newPinnedEndpoint = null;
+            StringCompiler sc = new StringCompiler();
+            sc.bind("pinned", newPinnedEndpoints);
+            sc.bind("this", this);
+            logger.debug(sc.c(
+                    "Endpoint {this.id} notified us that it has pinned"
+                            + " {pinned}."));
         }
 
-        // Check if that's different to what we think the pinned endpoint is.
+        pinnedEndpointsChanged(newPinnedEndpoints);
+    }
+
+    private void pinnedEndpointsChanged(List<String> pinnedEndpoints)
+    {
+        // Check if that's different to what we think the pinned endpoints are.
         boolean changed;
-        Endpoint oldPinnedEndpoint = this.getPinnedEndpoint();
         synchronized (pinnedEndpointSyncRoot)
         {
-            changed = newPinnedEndpoint != oldPinnedEndpoint;
-            if (changed)
+            changed = pinnedEndpoints.size() != this.pinnedEndpoints.size();
+            if (!changed)
             {
-                if (newPinnedEndpoint == null)
-                {
-                    this.weakPinnedEndpoint = null;
-                }
-                else
+                for (int i = 0; i < pinnedEndpoints.size(); i++)
                 {
-                    this.weakPinnedEndpoint
-                        = new WeakReference<>(newPinnedEndpoint);
+                    if (!pinnedEndpoints.get(i).
```

Comment:
I could be misunderstanding, but can pinnedEndpoints.get(i) be null? 

In our case, if you pin a stream then unpin the same stream (back to 0 pinnedEndpoints) this method will get called with a singleton list of [null].

It then seems like the resulting NPE here gets swallowed by someone and a Pinned Endpoint property event never fires.